apps/perl/5.18.2/man/man3/Socket.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Socket 3"
.TH Socket 3 "2013-10-27" "perl v5.18.2" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
"Socket" \- networking constants and support functions
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\f(CW\*(C`Socket\*(C'\fR a low-level module used by, among other things, the IO::Socket
family of modules. The following examples demonstrate some low-level uses but
a practical program would likely use the higher-level \s-1API\s0 provided by
\&\f(CW\*(C`IO::Socket\*(C'\fR or similar instead.
.PP
.Vb 1
\& use Socket qw(PF_INET SOCK_STREAM pack_sockaddr_in inet_aton);
\&
\& socket(my $socket, PF_INET, SOCK_STREAM, 0)
\&     or die "socket: $!";
\&
\& my $port = getservbyname "echo", "tcp";
\& connect($socket, pack_sockaddr_in($port, inet_aton("localhost")))
\&     or die "connect: $!";
\&
\& print $socket "Hello, world!\en";
\& print <$socket>;
.Ve
.PP
See also the \*(L"\s-1EXAMPLES\*(R"\s0 section.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This module provides a variety of constants, structure manipulators and other
functions related to socket-based networking. The values and functions
provided are useful when used in conjunction with Perl core functions such as
\&\fIsocket()\fR, \fIsetsockopt()\fR and \fIbind()\fR. It also provides several other support
functions, mostly for dealing with conversions of network addresses between
human-readable and native binary forms, and for hostname resolver operations.
.PP
Some constants and functions are exported by default by this module; but for
backward-compatibility any recently-added symbols are not exported by default
and must be requested explicitly. When an import list is provided to the
\&\f(CW\*(C`use Socket\*(C'\fR line, the default exports are not automatically imported. It is
therefore best practice to always to explicitly list all the symbols required.
.PP
Also, some common socket \*(L"newline\*(R" constants are provided: the constants
\&\f(CW\*(C`CR\*(C'\fR, \f(CW\*(C`LF\*(C'\fR, and \f(CW\*(C`CRLF\*(C'\fR, as well as \f(CW$CR\fR, \f(CW$LF\fR, and \f(CW$CRLF\fR, which map
to \f(CW\*(C`\e015\*(C'\fR, \f(CW\*(C`\e012\*(C'\fR, and \f(CW\*(C`\e015\e012\*(C'\fR. If you do not want to use the literal
characters in your programs, then use the constants provided here. They are
not exported by default, but can be imported individually, and with the
\&\f(CW\*(C`:crlf\*(C'\fR export tag:
.PP
.Vb 1
\& use Socket qw(:DEFAULT :crlf);
\&
\& $sock\->print("GET / HTTP/1.0$CRLF");
.Ve
.PP
The entire \fIgetaddrinfo()\fR subsystem can be exported using the tag \f(CW\*(C`:addrinfo\*(C'\fR;
this exports the \fIgetaddrinfo()\fR and \fIgetnameinfo()\fR functions, and all the
\&\f(CW\*(C`AI_*\*(C'\fR, \f(CW\*(C`NI_*\*(C'\fR, \f(CW\*(C`NIx_*\*(C'\fR and \f(CW\*(C`EAI_*\*(C'\fR constants.
.SH "CONSTANTS"
.IX Header "CONSTANTS"
In each of the following groups, there may be many more constants provided
than just the ones given as examples in the section heading. If the heading
ends \f(CW\*(C`...\*(C'\fR then this means there are likely more; the exact constants
provided will depend on the \s-1OS\s0 and headers found at compile-time.
.SS "\s-1PF_INET, PF_INET6, PF_UNIX, ...\s0"
.IX Subsection "PF_INET, PF_INET6, PF_UNIX, ..."
Protocol family constants to use as the first argument to \fIsocket()\fR or the
value of the \f(CW\*(C`SO_DOMAIN\*(C'\fR or \f(CW\*(C`SO_FAMILY\*(C'\fR socket option.
.SS "\s-1AF_INET, AF_INET6, AF_UNIX, ...\s0"
.IX Subsection "AF_INET, AF_INET6, AF_UNIX, ..."
Address family constants used by the socket address structures, to pass to
such functions as \fIinet_pton()\fR or \fIgetaddrinfo()\fR, or are returned by such
functions as \fIsockaddr_family()\fR.
.SS "\s-1SOCK_STREAM, SOCK_DGRAM, SOCK_RAW, ...\s0"
.IX Subsection "SOCK_STREAM, SOCK_DGRAM, SOCK_RAW, ..."
Socket type constants to use as the second argument to \fIsocket()\fR, or the value
of the \f(CW\*(C`SO_TYPE\*(C'\fR socket option.
.SS "\s-1SOCK_NONBLOCK. SOCK_CLOEXEC\s0"
.IX Subsection "SOCK_NONBLOCK. SOCK_CLOEXEC"
Linux-specific shortcuts to specify the \f(CW\*(C`O_NONBLOCK\*(C'\fR and \f(CW\*(C`FD_CLOEXEC\*(C'\fR flags
during a \f(CWsocket(2)\fR call.
.PP
.Vb 1
\& socket( my $sockh, PF_INET, SOCK_DGRAM|SOCK_NONBLOCK, 0 )
.Ve
.SS "\s-1SOL_SOCKET\s0"
.IX Subsection "SOL_SOCKET"
Socket option level constant for \fIsetsockopt()\fR and \fIgetsockopt()\fR.
.SS "\s-1SO_ACCEPTCONN, SO_BROADCAST, SO_ERROR, ...\s0"
.IX Subsection "SO_ACCEPTCONN, SO_BROADCAST, SO_ERROR, ..."
Socket option name constants for \fIsetsockopt()\fR and \fIgetsockopt()\fR at the
\&\f(CW\*(C`SOL_SOCKET\*(C'\fR level.
.SS "\s-1IP_OPTIONS, IP_TOS, IP_TTL, ...\s0"
.IX Subsection "IP_OPTIONS, IP_TOS, IP_TTL, ..."
Socket option name constants for IPv4 socket options at the \f(CW\*(C`IPPROTO_IP\*(C'\fR
level.
.SS "\s-1IPTOS_LOWDELAY, IPTOS_THROUGHPUT, IPTOS_RELIABILITY, ...\s0"
.IX Subsection "IPTOS_LOWDELAY, IPTOS_THROUGHPUT, IPTOS_RELIABILITY, ..."
Socket option value constants for \f(CW\*(C`IP_TOS\*(C'\fR socket option.
.SS "\s-1MSG_BCAST, MSG_OOB, MSG_TRUNC, ...\s0"
.IX Subsection "MSG_BCAST, MSG_OOB, MSG_TRUNC, ..."
Message flag constants for \fIsend()\fR and \fIrecv()\fR.
.SS "\s-1SHUT_RD, SHUT_RDWR, SHUT_WR\s0"
.IX Subsection "SHUT_RD, SHUT_RDWR, SHUT_WR"
Direction constants for \fIshutdown()\fR.
.SS "\s-1INADDR_ANY, INADDR_BROADCAST, INADDR_LOOPBACK, INADDR_NONE\s0"
.IX Subsection "INADDR_ANY, INADDR_BROADCAST, INADDR_LOOPBACK, INADDR_NONE"
Constants giving the special \f(CW\*(C`AF_INET\*(C'\fR addresses for wildcard, broadcast,
local loopback, and invalid addresses.
.PP
Normally equivalent to inet_aton('0.0.0.0'), inet_aton('255.255.255.255'),
inet_aton('localhost') and inet_aton('255.255.255.255') respectively.
.SS "\s-1IPPROTO_IP, IPPROTO_IPV6, IPPROTO_TCP, ...\s0"
.IX Subsection "IPPROTO_IP, IPPROTO_IPV6, IPPROTO_TCP, ..."
\&\s-1IP\s0 protocol constants to use as the third argument to \fIsocket()\fR, the level
argument to \fIgetsockopt()\fR or \fIsetsockopt()\fR, or the value of the \f(CW\*(C`SO_PROTOCOL\*(C'\fR
socket option.
.SS "\s-1TCP_CORK, TCP_KEEPALIVE, TCP_NODELAY, ...\s0"
.IX Subsection "TCP_CORK, TCP_KEEPALIVE, TCP_NODELAY, ..."
Socket option name constants for \s-1TCP\s0 socket options at the \f(CW\*(C`IPPROTO_TCP\*(C'\fR
level.
.SS "\s-1IN6ADDR_ANY, IN6ADDR_LOOPBACK\s0"
.IX Subsection "IN6ADDR_ANY, IN6ADDR_LOOPBACK"
Constants giving the special \f(CW\*(C`AF_INET6\*(C'\fR addresses for wildcard and local
loopback.
.PP
Normally equivalent to inet_pton(\s-1AF_INET6, \*(L"::\*(R"\s0) and
inet_pton(\s-1AF_INET6, \*(L"::1\*(R"\s0) respectively.
.SS "\s-1IPV6_ADD_MEMBERSHIP, IPV6_MTU, IPV6_V6ONLY, ...\s0"
.IX Subsection "IPV6_ADD_MEMBERSHIP, IPV6_MTU, IPV6_V6ONLY, ..."
Socket option name constants for IPv6 socket options at the \f(CW\*(C`IPPROTO_IPV6\*(C'\fR
level.
.SH "STRUCTURE MANIPULATORS"
.IX Header "STRUCTURE MANIPULATORS"
The following functions convert between lists of Perl values and packed binary
strings representing structures.
.ie n .SS "$family = sockaddr_family $sockaddr"
.el .SS "\f(CW$family\fP = sockaddr_family \f(CW$sockaddr\fP"
.IX Subsection "$family = sockaddr_family $sockaddr"
Takes a packed socket address (as returned by \fIpack_sockaddr_in()\fR,
\&\fIpack_sockaddr_un()\fR or the perl builtin functions \fIgetsockname()\fR and
\&\fIgetpeername()\fR). Returns the address family tag. This will be one of the
\&\f(CW\*(C`AF_*\*(C'\fR constants, such as \f(CW\*(C`AF_INET\*(C'\fR for a \f(CW\*(C`sockaddr_in\*(C'\fR addresses or
\&\f(CW\*(C`AF_UNIX\*(C'\fR for a \f(CW\*(C`sockaddr_un\*(C'\fR. It can be used to figure out what unpack to
use for a sockaddr of unknown type.
.ie n .SS "$sockaddr = pack_sockaddr_in $port, $ip_address"
.el .SS "\f(CW$sockaddr\fP = pack_sockaddr_in \f(CW$port\fP, \f(CW$ip_address\fP"
.IX Subsection "$sockaddr = pack_sockaddr_in $port, $ip_address"
Takes two arguments, a port number and an opaque string (as returned by
\&\fIinet_aton()\fR, or a v\-string). Returns the \f(CW\*(C`sockaddr_in\*(C'\fR structure with those
arguments packed in and \f(CW\*(C`AF_INET\*(C'\fR filled in. For Internet domain sockets,
this structure is normally what you need for the arguments in \fIbind()\fR,
\&\fIconnect()\fR, and \fIsend()\fR.
.ie n .SS "($port, $ip_address) = unpack_sockaddr_in $sockaddr"
.el .SS "($port, \f(CW$ip_address\fP) = unpack_sockaddr_in \f(CW$sockaddr\fP"
.IX Subsection "($port, $ip_address) = unpack_sockaddr_in $sockaddr"
Takes a \f(CW\*(C`sockaddr_in\*(C'\fR structure (as returned by \fIpack_sockaddr_in()\fR,
\&\fIgetpeername()\fR or \fIrecv()\fR). Returns a list of two elements: the port and an
opaque string representing the \s-1IP\s0 address (you can use \fIinet_ntoa()\fR to convert
the address to the four-dotted numeric format). Will croak if the structure
does not represent an \f(CW\*(C`AF_INET\*(C'\fR address.
.PP
In scalar context will return just the \s-1IP\s0 address.
.ie n .SS "$sockaddr = sockaddr_in $port, $ip_address"
.el .SS "\f(CW$sockaddr\fP = sockaddr_in \f(CW$port\fP, \f(CW$ip_address\fP"
.IX Subsection "$sockaddr = sockaddr_in $port, $ip_address"
.ie n .SS "($port, $ip_address) = sockaddr_in $sockaddr"
.el .SS "($port, \f(CW$ip_address\fP) = sockaddr_in \f(CW$sockaddr\fP"
.IX Subsection "($port, $ip_address) = sockaddr_in $sockaddr"
A wrapper of \fIpack_sockaddr_in()\fR or \fIunpack_sockaddr_in()\fR. In list context,
unpacks its argument and returns a list consisting of the port and \s-1IP\s0 address.
In scalar context, packs its port and \s-1IP\s0 address arguments as a \f(CW\*(C`sockaddr_in\*(C'\fR
and returns it.
.PP
Provided largely for legacy compatibility; it is better to use
\&\fIpack_sockaddr_in()\fR or \fIunpack_sockaddr_in()\fR explicitly.
.ie n .SS "$sockaddr = pack_sockaddr_in6 $port, $ip6_address, [$scope_id, [$flowinfo]]"
.el .SS "\f(CW$sockaddr\fP = pack_sockaddr_in6 \f(CW$port\fP, \f(CW$ip6_address\fP, [$scope_id, [$flowinfo]]"
.IX Subsection "$sockaddr = pack_sockaddr_in6 $port, $ip6_address, [$scope_id, [$flowinfo]]"
Takes two to four arguments, a port number, an opaque string (as returned by
\&\fIinet_pton()\fR), optionally a scope \s-1ID\s0 number, and optionally a flow label
number. Returns the \f(CW\*(C`sockaddr_in6\*(C'\fR structure with those arguments packed in
and \f(CW\*(C`AF_INET6\*(C'\fR filled in. IPv6 equivalent of \fIpack_sockaddr_in()\fR.
.ie n .SS "($port, $ip6_address, $scope_id, $flowinfo) = unpack_sockaddr_in6 $sockaddr"
.el .SS "($port, \f(CW$ip6_address\fP, \f(CW$scope_id\fP, \f(CW$flowinfo\fP) = unpack_sockaddr_in6 \f(CW$sockaddr\fP"
.IX Subsection "($port, $ip6_address, $scope_id, $flowinfo) = unpack_sockaddr_in6 $sockaddr"
Takes a \f(CW\*(C`sockaddr_in6\*(C'\fR structure. Returns a list of four elements: the port
number, an opaque string representing the IPv6 address, the scope \s-1ID,\s0 and the
flow label. (You can use \fIinet_ntop()\fR to convert the address to the usual
string format). Will croak if the structure does not represent an \f(CW\*(C`AF_INET6\*(C'\fR
address.
.PP
In scalar context will return just the \s-1IP\s0 address.
.ie n .SS "$sockaddr = sockaddr_in6 $port, $ip6_address, [$scope_id, [$flowinfo]]"
.el .SS "\f(CW$sockaddr\fP = sockaddr_in6 \f(CW$port\fP, \f(CW$ip6_address\fP, [$scope_id, [$flowinfo]]"
.IX Subsection "$sockaddr = sockaddr_in6 $port, $ip6_address, [$scope_id, [$flowinfo]]"
.ie n .SS "($port, $ip6_address, $scope_id, $flowinfo) = sockaddr_in6 $sockaddr"
.el .SS "($port, \f(CW$ip6_address\fP, \f(CW$scope_id\fP, \f(CW$flowinfo\fP) = sockaddr_in6 \f(CW$sockaddr\fP"
.IX Subsection "($port, $ip6_address, $scope_id, $flowinfo) = sockaddr_in6 $sockaddr"
A wrapper of \fIpack_sockaddr_in6()\fR or \fIunpack_sockaddr_in6()\fR. In list context,
unpacks its argument according to \fIunpack_sockaddr_in6()\fR. In scalar context,
packs its arguments according to \fIpack_sockaddr_in6()\fR.
.PP
Provided largely for legacy compatibility; it is better to use
\&\fIpack_sockaddr_in6()\fR or \fIunpack_sockaddr_in6()\fR explicitly.
.ie n .SS "$sockaddr = pack_sockaddr_un $path"
.el .SS "\f(CW$sockaddr\fP = pack_sockaddr_un \f(CW$path\fP"
.IX Subsection "$sockaddr = pack_sockaddr_un $path"
Takes one argument, a pathname. Returns the \f(CW\*(C`sockaddr_un\*(C'\fR structure with that
path packed in with \f(CW\*(C`AF_UNIX\*(C'\fR filled in. For \f(CW\*(C`PF_UNIX\*(C'\fR sockets, this
structure is normally what you need for the arguments in \fIbind()\fR, \fIconnect()\fR,
and \fIsend()\fR.
.ie n .SS "($path) = unpack_sockaddr_un $sockaddr"
.el .SS "($path) = unpack_sockaddr_un \f(CW$sockaddr\fP"
.IX Subsection "($path) = unpack_sockaddr_un $sockaddr"
Takes a \f(CW\*(C`sockaddr_un\*(C'\fR structure (as returned by \fIpack_sockaddr_un()\fR,
\&\fIgetpeername()\fR or \fIrecv()\fR). Returns a list of one element: the pathname. Will
croak if the structure does not represent an \f(CW\*(C`AF_UNIX\*(C'\fR address.
.ie n .SS "$sockaddr = sockaddr_un $path"
.el .SS "\f(CW$sockaddr\fP = sockaddr_un \f(CW$path\fP"
.IX Subsection "$sockaddr = sockaddr_un $path"
.ie n .SS "($path) = sockaddr_un $sockaddr"
.el .SS "($path) = sockaddr_un \f(CW$sockaddr\fP"
.IX Subsection "($path) = sockaddr_un $sockaddr"
A wrapper of \fIpack_sockaddr_un()\fR or \fIunpack_sockaddr_un()\fR. In a list context,
unpacks its argument and returns a list consisting of the pathname. In a
scalar context, packs its pathname as a \f(CW\*(C`sockaddr_un\*(C'\fR and returns it.
.PP
Provided largely for legacy compatibility; it is better to use
\&\fIpack_sockaddr_un()\fR or \fIunpack_sockaddr_un()\fR explicitly.
.PP
These are only supported if your system has <\fIsys/un.h\fR>.
.ie n .SS "$ip_mreq = pack_ip_mreq $multiaddr, $interface"
.el .SS "\f(CW$ip_mreq\fP = pack_ip_mreq \f(CW$multiaddr\fP, \f(CW$interface\fP"
.IX Subsection "$ip_mreq = pack_ip_mreq $multiaddr, $interface"
Takes an IPv4 multicast address and optionally an interface address (or
\&\f(CW\*(C`INADDR_ANY\*(C'\fR). Returns the \f(CW\*(C`ip_mreq\*(C'\fR structure with those arguments packed
in. Suitable for use with the \f(CW\*(C`IP_ADD_MEMBERSHIP\*(C'\fR and \f(CW\*(C`IP_DROP_MEMBERSHIP\*(C'\fR
sockopts.
.ie n .SS "($multiaddr, $interface) = unpack_ip_mreq $ip_mreq"
.el .SS "($multiaddr, \f(CW$interface\fP) = unpack_ip_mreq \f(CW$ip_mreq\fP"
.IX Subsection "($multiaddr, $interface) = unpack_ip_mreq $ip_mreq"
Takes an \f(CW\*(C`ip_mreq\*(C'\fR structure. Returns a list of two elements; the IPv4
multicast address and interface address.
.ie n .SS "$ip_mreq_source = pack_ip_mreq_source $multiaddr, $source, $interface"
.el .SS "\f(CW$ip_mreq_source\fP = pack_ip_mreq_source \f(CW$multiaddr\fP, \f(CW$source\fP, \f(CW$interface\fP"
.IX Subsection "$ip_mreq_source = pack_ip_mreq_source $multiaddr, $source, $interface"
Takes an IPv4 multicast address, source address, and optionally an interface
address (or \f(CW\*(C`INADDR_ANY\*(C'\fR). Returns the \f(CW\*(C`ip_mreq_source\*(C'\fR structure with those
arguments packed in. Suitable for use with the \f(CW\*(C`IP_ADD_SOURCE_MEMBERSHIP\*(C'\fR
and \f(CW\*(C`IP_DROP_SOURCE_MEMBERSHIP\*(C'\fR sockopts.
.ie n .SS "($multiaddr, $source, $interface) = unpack_ip_mreq_source $ip_mreq"
.el .SS "($multiaddr, \f(CW$source\fP, \f(CW$interface\fP) = unpack_ip_mreq_source \f(CW$ip_mreq\fP"
.IX Subsection "($multiaddr, $source, $interface) = unpack_ip_mreq_source $ip_mreq"
Takes an \f(CW\*(C`ip_mreq_source\*(C'\fR structure. Returns a list of three elements; the
IPv4 multicast address, source address and interface address.
.ie n .SS "$ipv6_mreq = pack_ipv6_mreq $multiaddr6, $ifindex"
.el .SS "\f(CW$ipv6_mreq\fP = pack_ipv6_mreq \f(CW$multiaddr6\fP, \f(CW$ifindex\fP"
.IX Subsection "$ipv6_mreq = pack_ipv6_mreq $multiaddr6, $ifindex"
Takes an IPv6 multicast address and an interface number. Returns the
\&\f(CW\*(C`ipv6_mreq\*(C'\fR structure with those arguments packed in. Suitable for use with
the \f(CW\*(C`IPV6_ADD_MEMBERSHIP\*(C'\fR and \f(CW\*(C`IPV6_DROP_MEMBERSHIP\*(C'\fR sockopts.
.ie n .SS "($multiaddr6, $ifindex) = unpack_ipv6_mreq $ipv6_mreq"
.el .SS "($multiaddr6, \f(CW$ifindex\fP) = unpack_ipv6_mreq \f(CW$ipv6_mreq\fP"
.IX Subsection "($multiaddr6, $ifindex) = unpack_ipv6_mreq $ipv6_mreq"
Takes an \f(CW\*(C`ipv6_mreq\*(C'\fR structure. Returns a list of two elements; the IPv6
address and an interface number.
.SH "FUNCTIONS"
.IX Header "FUNCTIONS"
.ie n .SS "$ip_address = inet_aton $string"
.el .SS "\f(CW$ip_address\fP = inet_aton \f(CW$string\fP"
.IX Subsection "$ip_address = inet_aton $string"
Takes a string giving the name of a host, or a textual representation of an \s-1IP\s0
address and translates that to an packed binary address structure suitable to
pass to \fIpack_sockaddr_in()\fR. If passed a hostname that cannot be resolved,
returns \f(CW\*(C`undef\*(C'\fR. For multi-homed hosts (hosts with more than one address),
the first address found is returned.
.PP
For portability do not assume that the result of \fIinet_aton()\fR is 32 bits wide,
in other words, that it would contain only the IPv4 address in network order.
.PP

apps/perl/5.18.2/lib/5.18.2/lib/perl5/x86_64-linux.bkp/Coro/Socket.pm
=head1 NAME

Coro::Socket - non-blocking socket-I/O

=head1 SYNOPSIS

 use Coro::Socket;

 # listen on an ipv4 socket
 my $socket = new Coro::Socket PeerHost => "localhost",
                               PeerPort => 'finger';

 # listen on any other type of socket
 my $socket = Coro::Socket->new_from_fh
                 (IO::Socket::UNIX->new
                     Local  => "/tmp/socket",
                     Type   => SOCK_STREAM,
                 );

=head1 DESCRIPTION

This module is an L<AnyEvent> user, you need to make sure that you use and
run a supported event loop.

This module implements socket-handles in a coroutine-compatible way,
that is, other coroutines can run while reads or writes block on the
handle. See L<Coro::Handle>, especially the note about prefering method
calls.

=head1 IPV6 WARNING

This module was written to imitate the L<IO::Socket::INET> API, and derive
from it. Since IO::Socket::INET does not support IPv6, this module does
neither.

Therefore it is not recommended to use Coro::Socket in new code. Instead,
use L<AnyEvent::Socket> and L<Coro::Handle>, e.g.:

   use Coro;
   use Coro::Handle;
   use AnyEvent::Socket;

   # use tcp_connect from AnyEvent::Socket
   # and call Coro::Handle::unblock on it.

   tcp_connect "www.google.com", 80, Coro::rouse_cb;
   my $fh = unblock +(Coro::rouse_wait)[0];

   # now we have a perfectly thread-safe socket handle in $fh
   print $fh "GET / HTTP/1.0\015\012\015\012";
   local $/;
   print <$fh>;

Using C<AnyEvent::Socket::tcp_connect> gives you transparent IPv6,
multi-homing, SRV-record etc. support.

For listening sockets, use C<AnyEvent::Socket::tcp_server>.

=over 4

=cut

package Coro::Socket;

use common::sense;

use Errno ();
use Carp qw(croak);
use Socket;
use IO::Socket::INET ();

use Coro::Util ();

use base qw(Coro::Handle IO::Socket::INET);

our $VERSION = 6.37;

our (%_proto, %_port);

sub _proto($) {
   $_proto{$_[0]} ||= do {
      ((getprotobyname $_[0])[2] || (getprotobynumber $_[0])[2])
         or croak "unsupported protocol: $_[0]";
   };
}

sub _port($$) {
   $_port{$_[0],$_[1]} ||= do {
      return $_[0] if $_[0] =~ /^\d+$/;

      $_[0] =~ /([^(]+)\s*(?:\((\d+)\))?/x
         or croak "unparsable port number: $_[0]";
      ((getservbyname $1, $_[1])[2]
        || (getservbyport $1, $_[1])[2]
        || $2)
         or croak "unknown port: $_[0]";
   };
}

sub _sa($$$) {
   my ($host, $port, $proto) = @_;

   $port or $host =~ s/:([^:]+)$// and $port = $1;

   my $_proto = _proto($proto);
   my $_port = _port($port, $proto);

   my $_host = Coro::Util::inet_aton $host
      or croak "$host: unable to resolve";

   pack_sockaddr_in $_port, $_host
}

=item $fh = new Coro::Socket param => value, ...

Create a new non-blocking tcp handle and connect to the given host
and port. The parameter names and values are mostly the same as for
IO::Socket::INET (as ugly as I think they are).

The parameters officially supported currently are: C<ReuseAddr>,
C<LocalPort>, C<LocalHost>, C<PeerPort>, C<PeerHost>, C<Listen>, C<Timeout>,
C<SO_RCVBUF>, C<SO_SNDBUF>.

   $fh = new Coro::Socket PeerHost => "localhost", PeerPort => 'finger';

=cut

sub _prepare_socket {
   my ($self, $arg) = @_;

   $self
}
   
sub new {
   my ($class, %arg) = @_;

   $arg{Proto}     ||= 'tcp';
   $arg{LocalHost} ||= delete $arg{LocalAddr};
   $arg{PeerHost}  ||= delete $arg{PeerAddr};
   defined ($arg{Type}) or $arg{Type} = $arg{Proto} eq "tcp" ? SOCK_STREAM : SOCK_DGRAM;

   socket my $fh, PF_INET, $arg{Type}, _proto ($arg{Proto})
      or return;

   my $self = bless Coro::Handle->new_from_fh (
      $fh,
      timeout       => $arg{Timeout},
      forward_class => $arg{forward_class},
      partial       => $arg{partial},
   ), $class
      or return;

   $self->configure (\%arg)
}

sub configure {
   my ($self, $arg) = @_;

   if ($arg->{ReuseAddr}) {
      $self->setsockopt (SOL_SOCKET, SO_REUSEADDR, 1)
         or croak "setsockopt(SO_REUSEADDR): $!";
   }

   if ($arg->{ReusePort}) {
      $self->setsockopt (SOL_SOCKET, SO_REUSEPORT, 1)
         or croak "setsockopt(SO_REUSEPORT): $!";
   }

   if ($arg->{Broadcast}) {
      $self->setsockopt (SOL_SOCKET, SO_BROADCAST, 1)
         or croak "setsockopt(SO_BROADCAST): $!";
   }

   if ($arg->{SO_RCVBUF}) {
      $self->setsockopt (SOL_SOCKET, SO_RCVBUF, $arg->{SO_RCVBUF})
         or croak "setsockopt(SO_RCVBUF): $!";
   }

   if ($arg->{SO_SNDBUF}) {
      $self->setsockopt (SOL_SOCKET, SO_SNDBUF, $arg->{SO_SNDBUF})
         or croak "setsockopt(SO_SNDBUF): $!";
   }

   if ($arg->{LocalPort} || $arg->{LocalHost}) {
      my @sa = _sa($arg->{LocalHost} || "0.0.0.0", $arg->{LocalPort} || 0, $arg->{Proto});
      $self->bind ($sa[0])
         or croak "bind($arg->{LocalHost}:$arg->{LocalPort}): $!";
   }

   if ($arg->{PeerHost}) {
      my @sa = _sa ($arg->{PeerHost}, $arg->{PeerPort}, $arg->{Proto});

      for (@sa) {
         $! = 0;

         if ($self->connect ($_)) {
            next unless writable $self;
            $! = unpack "i", $self->getsockopt (SOL_SOCKET, SO_ERROR);
         }

         $! or last;

         $!{ECONNREFUSED} or $!{ENETUNREACH} or $!{ETIMEDOUT} or $!{EHOSTUNREACH}
            or return;
      }
   } elsif (exists $arg->{Listen}) {
      $self->listen ($arg->{Listen})
         or return;
   }

   $self
}

1;

=back

=head1 AUTHOR

 Marc Lehmann <schmorp@schmorp.de>
 http://home.schmorp.de/

=cut

